fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ViewAnalyzer.java

```diff
@@ -28,23 +27,28 @@ public static void analyze(BaseViewStmt stmt, ConnectContext session) {
         Analyzer.analyze(stmt.getQueryStatement(), session);
         QueryRelation queryRelation = stmt.getQueryStatement().getQueryRelation();
 
-        List<Column> viewColumns = Lists.newArrayList();
+        // We create view columns from output fields of queryRelation if user-specified view columns are absent.
+        // in such cases, field.name and field.type instead of field.outputExpression.type are used to construct the
+        // columns of view's schema, since when queryRelation is union all stmt, field.outputExpression
+        // is always the leftmost child of the UnionOperator, so if one of the left most child's output column
+        // is const NULL and the corresponding output columns of the remaining children of UnionOperator
+        // are not NULL(i.e. VARCHAR), then field.type(VARCHAR) is inconsistent with the
+        // field.outputExpression.type(NULL) and the latter is incorrect. Wrong plans would be produced
+        // from such views, so we always adopts of field.type instead of field.outputExpression.type
+        List<Column> viewColumns = queryRelation.getScope().getRelationFields().getAllFields().stream()
+                .map(f -> new Column(f.getName(), f.getType())).collect(Collectors.toList());
```

Comment:
@HangyuanLiu getAllFields is equals with outputExpression?
@satanson don't merge before hangyuan review

Reply:
I have communicated with @satanson, getAllFields is not equals with outputExpression. That's exactly what this bug is about.
outputExpression is the leftmost child of the entire setOp tree. It is a grammatical concept, because the output expression of union is determined according to the output of the leftmost child. The field in Scope is the concept at the analyzer level, and the common type of all unions has been determined in the Analyzer